### PR TITLE
Handle 2D blocking with foreach

### PR DESCRIPTION
**Pull request by mlazos**
Previously blocking in foreach ops was only 1D. This PR allows handling kernels with 2D blocking with foreach as well.

Code when at least one dim matches:
[example code + output](https://gist.github.com/mlazos/9f100b21cfe2540f0a24303a8349c196)

Code when neither X or Y dim matches:
[example code + output](https://gist.github.com/mlazos/14e2a455f635896dface09be601595dd)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
talked via chat about this, should require the non-foreach dimension to have exactly same numel.

File: torch/_inductor/codegen/triton_foreach.py [hunk 1]
```diff
@@ -1,81 +1,140 @@
...
 
-        return partitions
+        return all_partitions
 
     def __init__(self):
         super().__init__()
-        self.block_size = 1024  # Try tuning this value
+        self.blocking_2d = False
+        self.block_size_1d = 1024  # Try tuning this value
+        self.block_size_2d = 32
```
Comment (jansel) [inline torch/_inductor/codegen/triton_foreach.py:82]: could we use the same autotuning we do for other kernels for foreach?  I worry 32 won't be the right block size.

**Review by jansel — COMMENTED** [review]
@shunting314 would you mind reviewing this one?  I am curious if there will be any interaction with #107902 which swaps x/y pids.

Any perf numbers?

Can you also include output code in the case where we can't combine them?

File: torch/_inductor/codegen/triton_foreach.py [hunk 2]
```diff
@@ -1,81 +1,140 @@
...
-                cur_count = read_write_count
+
+            if tiled_groups[1] == 1:
+                ForeachKernel._update_partition(
+                    partition_state_1d, read_write_count, node_info
+                )
             else:
-                cur_count += read_write_count
-                cur_partition.append(node)
+                y_elem = tiled_groups[0]
```
Comment (shunting314) [inline torch/_inductor/codegen/triton_foreach.py:64]: So if the kernel is a reduction kernel, `tiled_groups[0]` will be numel and `tiled_groups[1]` will be rnumel. In this case `tiled_groups[0]` would be actually X-numel. Assigning it to `y_elem` looks a bit confusing here. Although I need understand more how we use `y_elem` in other places.

Comment (shunting314) [reply torch/_inductor/codegen/triton_foreach.py:64]: I guess foreach kernel does not handle reduction? If that's the case, let's add an assertion for 'rnumel == 1' above

File: torch/_inductor/codegen/triton_foreach.py (hunk 1 above)
Comment (mlazos) [reply torch/_inductor/codegen/triton_foreach.py:82]: Yeah I think we can, do I just add it to the meta params?

File: torch/_inductor/codegen/triton_foreach.py (hunk 2 above)
Comment (mlazos) [reply torch/_inductor/codegen/triton_foreach.py:64]: Currently I don't lower any reduction kernels, but was planning to  in the future, how should I handle this consistently? or should I just add an if checking if it's a reduction?

File: torch/_inductor/codegen/triton_foreach.py [hunk 3]
```diff
@@ -123,14 +184,26 @@ def codegen_kernel(self, name=None):
...
-                self.codegen_pid_range(code, num_elems)
+                assert len(sub_kernel.numels) <= 3
+                # TODO mlazos: support dynamic shapes
+                numel_ind = 0 if not self.blocking_2d else 1
+                self.codegen_pid_range(code, int(sub_kernel.numels[numel_ind]))
                 with code.indent():
-                    code.splice(f"xnumel = {num_elems}")
+                    if self.blocking_2d:
+                        code.splice(f"ynumel = {sub_kernel.numels[0]}")
+                        code.splice(f"xnumel = {sub_kernel.numels[1]}")
```
Comment (shunting314) [inline torch/_inductor/codegen/triton_foreach.py:203]: This interpretation of x/y pids is consistent with https://github.com/pytorch/pytorch/pull/107902 . cc @jansel

File: torch/_inductor/codegen/triton_foreach.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/codegen/triton_foreach.py:82]: you might need to change the decorator

**Review by jansel — APPROVED** [review]

**Comment by mlazos** [comment]
Perf numbers:
https://gist.github.com/mlazos/31887fa37788f7ad29d0988b8872fa0e

**Comment by mlazos** [comment]
I'll add the tuning in a separate PR to unblock some other work

**Comment by ZainRizvi** [comment]
Fyi @mlazos, the force merge wasn't necessary. If the summary comment at the top calls out the unrelated jobs, normal merges will work just fine:

image

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this, but test_2d_blocking is failing in trunk, probably a landrace as PR was green' -c landrace

https://hud.pytorch.org/pytorch/pytorch/commit/f87ffe473d4825c15eaea0360baf08cad49979de

**Comment by mlazos** [comment]
Should be ready to go in after https://github.com/pytorch/pytorch/pull/107902#pullrequestreview-1596152846

**Pull request closed without merging**